src/webgpu/api/validation/vertex_state.spec.ts

```diff
@@ -303,73 +310,169 @@ g.test('vertex_attribute_shaderLocation_limit')
+   - Test for the potentially conflicting attributes in various places in the buffers (with dummy attributes)
+   - Test for various shaderLocations that conflict or not`
+  )
+  .subcases(() =>
+    params()
+      .combine(poptions('vertexBufferIndexA', [0, 1, kMaxVertexBuffers - 1]))
+      .combine(poptions('vertexBufferIndexB', [0, 1, kMaxVertexBuffers - 1]))
+      .combine(pbool('testAttributeAtStartA'))
+      .combine(pbool('testAttributeAtStartB'))
+      .combine(poptions('shaderLocationA', [0, 1, 7, kMaxVertexAttributes - 1]))
+      .combine(poptions('shaderLocationB', [0, 1, 7, kMaxVertexAttributes - 1]))
+      .combine(poptions('extraAttributes', [0, 4]))
+  )
+  .fn(t => {
+    const {
+      vertexBufferIndexA,
+      vertexBufferIndexB,
+      testAttributeAtStartA,
+      testAttributeAtStartB,
+      shaderLocationA,
+      shaderLocationB,
+      extraAttributes,
+    } = t.params;
+
+    const vertexBufferAttributes: GPUVertexAttributeDescriptor[][] = [];
+    vertexBufferAttributes[vertexBufferIndexA] = [];
+    vertexBufferAttributes[vertexBufferIndexB] = [];
+
+    // Add the dummy attributes for attribute A
+    const attributesA = vertexBufferAttributes[vertexBufferIndexA];
+    let currentLocation = 0;
+    for (let i = 0; i < extraAttributes; i++) {
+      if (currentLocation === shaderLocationA || currentLocation === shaderLocationB) {
+        currentLocation++;
+        continue;
+      }
+
+      attributesA.push({ format: 'float', shaderLocation: currentLocation, offset: 0 });
+      currentLocation++;
+    }
+
+    // Add attribute A
+    if (testAttributeAtStartA) {
+      attributesA.unshift({ format: 'float', offset: 0, shaderLocation: shaderLocationA });
+    } else {
+      attributesA.push({ format: 'float', offset: 0, shaderLocation: shaderLocationA });
+    }
+
+    // Add attribute B potentially in the same list of attributes as A
+    const attributesB = vertexBufferAttributes[vertexBufferIndexB];
+    if (testAttributeAtStartB) {
+      attributesB.unshift({ format: 'float', offset: 0, shaderLocation: shaderLocationB });
+    } else {
+      attributesB.push({ format: 'float', offset: 0, shaderLocation: shaderLocationB });
+    }
+
+    // Use the attributes to make the list of vertex buffers not that we might be setting the same vertex
+    // buffer twice but that only happens when it is the only vertex buffer.
+    const vertexBuffers: GPUVertexBufferLayoutDescriptor[] = [];
+    vertexBuffers[vertexBufferIndexA] = { arrayStride: 256, attributes: attributesA };
+    vertexBuffers[vertexBufferIndexB] = { arrayStride: 256, attributes: attributesB };
+
+    const success = shaderLocationA !== shaderLocationB;
+    t.testVertexState(success, { vertexBuffers });
+  });
+
+g.test('vertex_shader_input_location_limit')
+  .desc(
+    `Test that vertex shader's input's location decoration must be less than maxVertexAttributes.
+   - Test for shaderLocation 0, 1, limit - 1, limit`
```

Comment:
Maybe add some weird cases like `-1`, `2**32`? but more likely those will need to be wgsl tests